### PR TITLE
net module

### PR DESCRIPTION
**Pull request by aliib**
This pull request proposes a new electron module called the net module. It is essentially a JavaScript binding of Chromium native networking library (the net library) and provides an API similar to the Node.js HTTP module.

Currently, Electron apps developers are expected to use the native Node.js, or other helper libraries built on top of it, to issue their HTTP(S) requests. However, it is widely known that Node.js provides a quite minimal support for web proxies.

Using Chromium network stack offers substantial advantages over the HTTP/HTTPS implementation of Node.js, specifically regarding web proxies’ support. Following is a non-exhaustive list of such features:
- Automatic management of system proxy configuration, support of wpad protocol and proxy pac configuration files.
- Tunneling of HTTPS requests.
- Support for authenticating proxies using basic, digest, NTLM, Kerberos or negotiate authentication schemes.
- Traffic monitoring proxies: Fiddler-like proxies used for access control and monitoring.

While it is possible to provide support for most of the previously listed features on top of Node.js, eventually using additional npm packages, the task is not trivial and requires a substantial development effort and testing. Moreover, we are not aware of any public npm package that solves the authenticating proxies’ issue.

The initial design of Node.js as server-side platform explains while it has always lacked reasonable support for web proxies as the latters are typically deployed in client environments. 

Electron, being primarily targeted as a client-side framework, is expected to provide a robust networking API capable of automatically handling all issues introduced by usage of web proxies.

The present module solves all of the aforementioned problems. The API is quite simple and greatly mimics the familiar Node.j HTTP. So no substantial learning effort is required.

The implementation of the net module follows the usual binding patterns of electron. It extends the JavaScript context of the browser process by adding a new module. Applicative code not using the new module wouldn’t be impacted in any aspect.

On the other hand, applications willing to use the new module would automatically benefit from a robust and widely used networking layer targeted for client-side environments.

### CODE REVIEW

File: atom/browser/api/atom_api_url_request.cc [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+struct Converter<scoped_refptr<const net::HttpResponseHeaders>> {
+  static v8::Local<v8::Value> ToV8(
+    v8::Isolate* isolate,
+    scoped_refptr<const net::HttpResponseHeaders> val) {
+    mate::Dictionary dict = mate::Dictionary::CreateEmpty(isolate);
+    if (val) {
+      size_t iter = 0;
+      std::string name;
+      std::string value;
+      while (val->EnumerateHeaderLines(&iter, &name, &value)) {
```
Comment (deepak1556) [inline atom/browser/api/atom_api_url_request.cc]: Conversion should mimic https://github.com/electron/electron/blob/master/atom/browser/api/atom_api_web_contents.cc#L146 , maybe converter can be moved to `atom/commom/native_mate_converters/net_converter`

File: atom/browser/api/atom_api_url_request.cc [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+  std::string session_name;
+  dict.Get("session", &session_name);
+
+  auto session = Session::FromPartition(args->isolate(), session_name);
+
+  auto browser_context = session->browser_context();
+  auto api_url_request = new URLRequest(args->isolate(), args->GetThis());
+  auto weak_ptr = api_url_request->weak_ptr_factory_.GetWeakPtr();
+  auto atom_url_request = AtomURLRequest::Create(
+    browser_context,
```
Comment (deepak1556) [inline atom/browser/api/atom_api_url_request.cc]: 4 spaces. Same in other function call places.

File: atom/browser/net/atom_url_request.cc [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    base::WeakPtr<api::URLRequest> delegate) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  DCHECK(browser_context);
+  DCHECK(!url.empty());
+
+  auto request_context_getter = browser_context->url_request_context_getter();
+
+  DCHECK(request_context_getter);
+
+  auto context = request_context_getter->GetURLRequestContext();
```
Comment (deepak1556) [inline atom/browser/net/atom_url_request.cc]: Should not be called on UI thread.

File: atom/browser/net/atom_url_request.cc [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+  DCHECK(context);
+
+  scoped_refptr<AtomURLRequest> atom_url_request =
+    new AtomURLRequest(delegate);
+
+  atom_url_request->request_ = context->CreateRequest(GURL(url),
```
Comment (deepak1556) [inline atom/browser/net/atom_url_request.cc]: Moved to IO thread.

File: atom/browser/net/atom_url_request.cc [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+  content::BrowserThread::PostTask(
+    content::BrowserThread::IO,
+    FROM_HERE,
+    base::Bind(&AtomURLRequest::DoCancel, this));
+}
+
+void AtomURLRequest::SetExtraHeader(const std::string& name,
+                               const std::string& value) const {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  request_->SetExtraRequestHeaderByName(name, value, true);
```
Comment (deepak1556) [inline atom/browser/net/atom_url_request.cc]: Should be move to IO thread.

File: lib/browser/api/net.js [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+    })
+
+    // Set back and forward links.
+    this._url_request = urlRequest
+    urlRequest._request = this
+
+    // This is a copy of the extra headers structure held by the native
+    // net::URLRequest. The main reason is to keep the getHeader API synchronous
+    // after the request starts.
+    this._extra_headers = {}
```
Comment (deepak1556) [inline lib/browser/api/net.js]: `extraHeaders`

File: atom/browser/api/atom_api_url_request.cc (hunk 1 above)
Comment (aliib) [inline atom/browser/api/atom_api_url_request.cc]: ok, thanks for the review. We will all issues in the next commit

**Comment by deepak1556** [comment]
* Pending requests should be destroyed when its context is shutting down, otherwise app will crash in `net::URLRequestContext::AssertNoURLRequests` check.

To reproduce:

```
// main.js
const {app, net} = require('electron')
app.on('ready', () => {
  const req = net.request({url: 'https://google.com'})
})
```

`atom::URLRequest` should override `URLRequestContextGetterObserver::OnContextShuttingDown` and destroy any pending request.

* Net module should not be made available before `ready` event of `app` since no request context is available otherwise using the methods will lead to crash.

File: atom/browser/api/atom_api_url_request.cc [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+  EmitRequestEvent(true, "abort");
+
+  if (response_state_.Started() && !response_state_.Ended()) {
+    EmitResponseEvent(true, "aborted");
+  }
+  Close();
+}
+
+bool URLRequest::SetExtraHeader(const std::string& name,
+                           const std::string& value) {
```
Comment (deepak1556) [inline atom/browser/api/atom_api_url_request.cc]: alignment mismatch.

File: atom/browser/net/atom_url_request.cc [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+void AtomURLRequest::Cancel() const {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  content::BrowserThread::PostTask(
+      content::BrowserThread::IO,
+      FROM_HERE,
+      base::Bind(&AtomURLRequest::DoCancel, this));
+}
+
+void AtomURLRequest::SetExtraHeader(const std::string& name,
+                               const std::string& value) const {
```
Comment (deepak1556) [inline atom/browser/net/atom_url_request.cc]: alignment mismatch.

File: atom/browser/net/atom_url_request.cc [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+        base::Bind(&AtomURLRequest::DoCancelAuth, this));
+  } else {
+    content::BrowserThread::PostTask(
+        content::BrowserThread::IO, FROM_HERE,
+        base::Bind(&AtomURLRequest::DoSetAuth, this, username, password));
+  }
+}
+
+void AtomURLRequest::DoWriteBuffer(
+  scoped_refptr<const net::IOBufferWithSize> buffer,
```
Comment (deepak1556) [inline atom/browser/net/atom_url_request.cc]: 4 spaces.

File: atom/browser/net/atom_url_request.cc [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+  request_->SetAuth(net::AuthCredentials(username, password));
+}
+
+void AtomURLRequest::DoCancelAuth() const {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::IO);
+  request_->CancelAuth();
+}
+
+void AtomURLRequest::OnAuthRequired(net::URLRequest* request,
+                    net::AuthChallengeInfo* auth_info) {
```
Comment (deepak1556) [inline atom/browser/net/atom_url_request.cc]: alignment mismatch.

File: atom/browser/api/atom_api_url_request.cc (hunk 7 above)
Comment (deepak1556) [inline atom/browser/api/atom_api_url_request.cc]: can use [clang-format](https://github.com/electron/electron/blob/master/docs/development/clang-format.md)

**Comment by zcbenz** [comment]
This is something I always wanted to do but never tried, awesome work! 🎆

The PR is very large so the my review will take some time.

File: atom/browser/api/atom_api_url_request.h [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+  scoped_refptr<net::HttpResponseHeaders> RawResponseHeaders() const;
+  uint32_t ResponseHttpVersionMajor() const;
+  uint32_t ResponseHttpVersionMinor() const;
+
+  template <typename... ArgTypes>
+  std::array<v8::Local<v8::Value>, sizeof...(ArgTypes)> BuildArgsArray(
+      ArgTypes... args) const;
+
+  template <typename... ArgTypes>
+  void EmitRequestEvent(ArgTypes... args);
```
Comment (zcbenz) [inline atom/browser/api/atom_api_url_request.h]: How about moving this to the `.cc` file as a helper function?

And to make the code simpler (and also handle the micro tasks correctly), we can also refactor `event_emitter_caller.h` to make `CallEmitWithArgs` accept  a custom method:

```c++
template<typename... ArgTypes>
void CustomEmit(v8::Isolate* isolate,
                v8::Local<v8::Object> object,
                const ArgTypes&... args) {
  std::vector<v8::Local<v8::Value>> converted_args = {
    mate::ConvertToV8(isolate, args)...
  };
  mate::internal::CallEmitWithArgs(isolate, object, &converted_args);
}
```

**Comment by aliib** [comment]
thanks for the review.

File: atom/browser/api/atom_api_net.cc [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      .SetProperty("URLRequest", &Net::URLRequest)
+      .SetMethod("_RequestGarbageCollectionForTesting",
+                 &Net::RequestGarbageCollectionForTesting);
+}
+
+v8::Local<v8::Value> Net::URLRequest(v8::Isolate* isolate) {
+  return URLRequest::GetConstructor(isolate)->GetFunction();
+}
+
+void Net::RequestGarbageCollectionForTesting() {
```
Comment (zcbenz) [inline atom/browser/api/atom_api_net.cc]: The `v8Util` is a hidden module so you can use any meaningful name without worrying about adding underscore.

**Comment by davej** [comment]
Awesome. Am I correct to think this means requests using the `net` module will appear in the network panel of the Dev Tools?

It may be worthwhile to explicitly document the specific API differences between `net` and node's `http`/`https` modules. I imagine a lot of people will be looking to transition their code to `net` instead so it would be handy to know what API differences to look out for when migrating.

**Comment by aliib** [comment]
@davej 
I don't think we would be able to see `net` module requests in the network panel of a Dev Tools. 
`net` requests are not associated with any particular page while a Dev Tools page shows network requests for a particular page. If, at some point, Electron provides a Browser-process Dev Tools, then yes it would be feasible to inspect `net` module requests. However, contrary to Node.js requests, `net` module requests will automatically appear in usual traffic-inspection tools such as Fiddler.

**Comment by zcbenz** [comment]
👍

**Comment by MarshallOfSound** [comment]
@zeke Needs to be added to the docs seeds

**Pull request merged**